### ISSUE TITLE
ArrayBlockingQueueBuffer is activated unexpected in the IF_POSSIBLE mode

### ISSUE THREAD

**Issue by lijial** [apache/skywalking#6045]
Dear skywalking team, 
While reading skywalking agent code(version 8.3.0) and I notice in ArrayBlockingQueueBuffer.java, 
save() method, when strategy is IF_POSSIBLE, there only **queue.offer(data)** without catch clause .
As queue.offer() may throw an exception, this may interrupt TracingContext.stopSpan() upward.
Is it a bug?
Should I suggest to add catch clause for calling offer() ?

**Comment by wu-sheng** [comment]
I think if possibleb strategy is not using ArrayBlockingQueueBuffer. Could you recheck?

**Comment by lijial** [comment]
![image](https://user-images.githubusercontent.com/4213303/102773745-b5f02f80-43c4-11eb-8902-5a3b17a5d6ec.png)
In this pic above, as red box shows, so I still hold the doubt as topic

**Comment by wu-sheng** [comment]
Ha, I see. There are 2 separate things. 
1. From a tech perspective, there should be an NPE catch. If you want, this should be fixed. Welcome to send a pull request.
2. In the real world, this condition will never execute, take a look at here, https://github.com/apache/skywalking/blob/bf639d3ec6d3cad994d6a5234ae8420fe86021d5/apm-commons/apm-datacarrier/src/main/java/org/apache/skywalking/apm/commons/datacarrier/buffer/Channels.java#L38-L42.

**Comment by lijial** [comment]
@wu-sheng Dear
![image](https://user-images.githubusercontent.com/4213303/102777559-c3f57e80-43cb-11eb-88c4-72f8b36c761d.png)
https://github.com/apache/skywalking/blob/bf639d3ec6d3cad994d6a5234ae8420fe86021d5/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/TraceSegmentServiceClient.java#L63 to 69

After init in Channels.java as you mention above, then next, code will run to boot() method as pic, 
_carrier.setBufferStrategy(BufferStrategy.IF_POSSIBLE);_
so this line will be execute runtime
https://github.com/apache/skywalking/blob/bf639d3ec6d3cad994d6a5234ae8420fe86021d5/apm-commons/apm-datacarrier/src/main/java/org/apache/skywalking/apm/commons/datacarrier/buffer/ArrayBlockingQueueBuffer.java#L45

**Comment by wu-sheng** [comment]
Oh, this is a very serious bug we haven't detected for a very long time. Actually, the issue is, strategy determines the  channel implementation. In the `if_possible`, should be `Buffer` rather than `ArrayBlockingQueueBuffer`.

So, the change should be **remove** `DataCarrier#setBufferStrategy` method, and require `BufferStrategy` in the constructor.

Are you willing to fix this issue?

**Comment by wu-sheng** [comment]
Also, note, the exception(NPE) wouldn't happen, as in SkyWalking, no one should put a NULL into the queue, right?

**Comment by lijial** [comment]
> Oh, this is a very serious bug we haven't detected for a very long time. Actually, the issue is, strategy determines the channel implementation. In the `if_possible`, should be `Buffer` rather than `ArrayBlockingQueueBuffer`.
> 
> So, the change should be **remove** `DataCarrier#setBufferStrategy` method, and require `BufferStrategy` in the constructor.
> 
> Are you willing to fix this issue?

I'm willing to fix the issue, I would pull add commit a pr later, thank you for give me the chance

**Comment by lijial** [comment]
> Also, note, the exception(NPE) wouldn't happen, as in SkyWalking, no one should put a NULL into the queue, right?

Yes, that is right after checking it carefully

**Comment by lijial** [comment]
> Also, note, the exception(NPE) wouldn't happen, as in SkyWalking, no one should put a NULL into the queue, right?
I thought there's something wrong in the constructor(DataCarrierTest:testProduce() Line 60 ) .
And also DataCarrierTest:testBlockingProduce() method
https://github.com/apache/skywalking/commit/09c0e8e32d643eb8a4189225925389ab0e483252#diff-552816423049ee3be17a365cfcdc999a9af8f27bac814bcdcfb310faab82aa01L50

**Comment by wu-sheng** [comment]
Is my pull request resolving your concern?

**Comment by lijial** [comment]
> Is my pull request resolving your concern?

Yes, after compiling and testing , now it is fit expectation
And what I think code below would be delete, right?
https://github.com/apache/skywalking/blob/10772d388e12d2e79f999b2e084d9dd94771a56a/apm-commons/apm-datacarrier/src/main/java/org/apache/skywalking/apm/commons/datacarrier/buffer/ArrayBlockingQueueBuffer.java#L44-L45

**Comment by wu-sheng** [comment]
Logically yes. If you want, feel free to send another pull request to fix after #6048 merged

**Comment by lijial** [comment]
> Logically yes. If you want, feel free to send another pull request to fix after #6048 merged

OK, I would send another pull later

**Comment by wu-sheng** [comment]
Merged.
